src/estimators/stratified_sketch.py

```diff
@@ -0,0 +1,294 @@
+      that: the other StratifiedSketch to be merged.
+
+    Returns:
+      A merged StratifiedSketch from the input.
+    """
+    assert isinstance(this, StratifiedSketch)
+    this.assert_compatible(that)
+    max_freq = this.max_freq
+    max_key = str(max_freq) + '+'
+    merged_sketch = copy.deepcopy(this)
+
+    this_one_plus = this.cardinality_sketch_factory(0)
+    that_one_plus = that.cardinality_sketch_factory(0)
+    for k in range(1, max_freq):
+      this_one_plus = self.sketch_union(this_one_plus, this.sketches[k])
+      that_one_plus = self.sketch_union(that_one_plus, that.sketches[k])
+
+    this_one_plus = self.sketch_union(this_one_plus, this.sketches[max_key])
+    that_one_plus = self.sketch_union(that_one_plus, that.sketches[max_key])
+
+    for k in range(1, max_freq):
+      # Calculate A(k) & B(0) = A(k) - (A(k) & B(1+))
+      merged = self.sketch_difference(
+          this.sketches[k],
+          self.sketch_intersection(this.sketches[k], that_one_plus))
+
+      # Calculate A(0) & B(k) = B(k) - (B(k) & A(1+))
+      merged = self.sketch_union(
+          merged,
+          self.sketch_difference(
+              that.sketches[k],
+              self.sketch_intersection(this_one_plus, that.sketches[k])))
+
+      # Calculate A(i) & B(k-i)
+      for i in range(1, k):
+        merged = self.sketch_union(
+            merged,
+            self.sketch_intersection(this.sketches[i], that.sketches[(k - i)]))
+      merged_sketch.sketches[k] = merged
+
+    # Calculate Merged(max_freq)
+    merged = this.sketches[max_key]
+    rest = that_one_plus
+    for k in range(1, max_freq):
+      merged = self.sketch_union(
+          merged, self.sketch_intersection(this.sketches[max_freq - k], rest))
+      rest = self.sketch_difference(rest, that.sketches[k])
+
+    merged = self.sketch_union(
+        merged,
+        self.sketch_difference(
+            that.sketches[max_key],
+            self.sketch_intersection(that.sketches[max_key], this_one_plus)))
+    merged_sketch.sketches[max_key] = merged
+
+    # Calculate Merged(1+)
+    merged_one_plus = None
+    for k in range(1, max_freq):
+      merged_one_plus = self.sketch_union(merged_one_plus,
+                                          merged_sketch.sketches[k])
+      merged_one_plus = self.sketch_union(merged_one_plus,
+                                    merged_sketch.sketches[max_key])
+    merged_sketch.sketches[ONE_PLUS] = merged_one_plus
+    return merged_sketch
+
+  def estimate_cardinality(self, stratified_sketch):
+    """Estimate the cardinality of a StratifiedSketch.
+
+    Args:
+     stratified_sketch: a StratifiedSketch object.
+
+    Returns:
+      A dictionary: the key is the frequency and the value is the corresponding
+      cardinality.
+    """
+    result = {}
+    for freq, sketch in stratified_sketch.sketches.items():
+      result[freq] = self.cardinality_estimator([sketch])
+    return result
```

Comment:
This should return a cumulative histogram.  Also, self.cardinality_estimator([sketch]) itself returns a histogram, which makes me suspect that this is not being tested?

Reply:
It was being tested, was assuming a dict output though. Now we are outputting a cumulative histogram.